File: ljh/level2/뉴스 클러스터링/NewsClustering.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Pattern;

/**
 * 뉴스 클러스터링
 */
public class NewsClustering {
    public int solution(String str1, String str2) {
        str1 = str1.toUpperCase();
        str2 = str2.toUpperCase();

        List<String> a = divide(str1);
        List<String> b = divide(str2);

        Collections.sort(a);
        Collections.sort(b);

        int intersection = searchDuplication(a, b);
        int union = a.size() + b.size() - intersection;
        return (union == 0) ? 65536 : intersection * 65536 / union;
    }

    public List<String> divide(String str) {
        List<String> list = new ArrayList<>();
        String pattern = "^[A-Z]*$";

        int length = str.length();
        int last = length - 1;
        String sub;
        for (int i = 0; i < last; i++) {
            sub = str.substring(i, i + 2);
            if (Pattern.matches(pattern, sub)) {
                list.add(sub);
            }
        }

        return list;
    }

    private int searchDuplication(List<String> a, List<String> b) {
        int pointer = 0;
        int duplication = 0;
        for (int i = 0; i < a.size(); i++) {
            for (int j = pointer; j < b.size(); j++) {
                if (a.get(i).equals(b.get(j))) {
                    pointer = j + 1;
                    duplication++;
                    break;
                }
            }
        }
        return duplication;
    }
}

File: ljh/level2/뉴스 클러스터링/NewsClustering.js
// 뉴스 클러스터링
function solution(str1, str2) {
    str1 = str1.toUpperCase();
    str2 = str2.toUpperCase();
    
    const list1 = divide(str1);
    const list2 = divide(str2);  
    
    list1.sort();
    list2.sort();
    
    let pointer = 0;
    let intersection = 0;
    for (let i = 0; i < list1.length; i++) {
        for (let j = pointer; j < list2.length; j++) {
            if (list1[i] === list2[j]) {
                intersection++;
                pointer = j + 1;
                break;
            }
        }
    }

    const union = list1.length + list2.length - intersection;
    return union === 0? 65536 : Math.floor(intersection * 65536 / union);
}

function divide (str) {
    const list = [];
    for(let i = 0; i < str.length - 1; i++) {
        let char1 = str[i].charCodeAt();
        if(char1 < 65 || char1 > 90) {
            continue;
        }
        let char2 = str[i+1].charCodeAt();
        if(char2 < 65 || char2 > 90) {
            continue;
        }
        
        list.push(String.fromCharCode(char1, char2));
    }
    return list;
}